### PR TITLE
test: expanded storage integration tests

### PR DESCRIPTION
**Pull request by jmdeal**
Fixes #N/A 

**Description**
This PR expands upon the existing storage integration tests in light of https://github.com/kubernetes-sigs/karpenter/pull/783.

**How was this change tested?**
`go test` and `/karpenter snapshot`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jmdeal — COMMENTED** [review]
/karpenter snapshot

File: test/suites/integration/storage_test.go [hunk 1]
```diff
@@ -16,104 +16,212 @@ package integration_test
...
+			env.ExpectCreatedNodeCount("==", 1)
 		})
-
-		pvc := test.PersistentVolumeClaim(test.PersistentVolumeClaimOptions{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "ebs-claim",
-			},
-			StorageClassName: aws.String(storageClassName),
-			Resources:        v1.ResourceRequirements{Requests: v1.ResourceList{v1.ResourceStorage: resource.MustParse("5Gi")}},
+		It("should run a pod with a pre-bound persistent volume (non-existent storage class)", func() {
```
Comment (jonathan-innis) [inline test/suites/integration/storage_test.go:57]: Should we test the case with us safely handling the case where someone is using a dynamically created PVC with a non-existent storage class? I know we discussed this edge-case and it might be nice to have some handling for it in our tests. Also happy to punt on this one and save it for another PR but it's good to know that this is a possibility

File: test/suites/integration/storage_test.go [hunk 2]
```diff
@@ -16,104 +16,212 @@ package integration_test
...
+					Name: pvc.Spec.VolumeName,
+				},
+				StorageClassName: "non-existent-storage-class",
+				Zones:            []string{shuffledAZs[0]},
+			})
+			pod := test.Pod(test.PodOptions{
+				PersistentVolumeClaims: []string{pvc.Name},
+			})
+			env.ExpectCreated(nodeClass, nodePool, pv, pvc, pod)
+			env.EventuallyExpectHealthy(pod)
```
Comment (jonathan-innis) [inline test/suites/integration/storage_test.go:93]: So in this case, should we unschedule the pod, then create a new pod and expect that pod to be healthy, use the PV, and schedule to the same zone?

File: test/suites/integration/storage_test.go [hunk 3]
```diff
@@ -16,104 +16,212 @@ package integration_test
...
+					StorageClassName: &storageClass.Name,
+				})
+			})
+			pods := lo.Map(pvcs, func(pvc *v1.PersistentVolumeClaim, _ int) *v1.Pod {
+				return test.Pod(test.PodOptions{
+					PersistentVolumeClaims: []string{pvc.Name},
+				})
+			})
+			env.ExpectCreated(nodeClass, nodePool, storageClass, pvcs[0], pods[0])
+			env.EventuallyExpectHealthy(pods[0])
```
Comment (jonathan-innis) [inline test/suites/integration/storage_test.go:173]: Consider a comment here on why we have this ordering

File: test/suites/integration/storage_test.go [hunk 4]
```diff
@@ -16,104 +16,212 @@ package integration_test
...
+				}
+			}
+			storageClass = test.StorageClass(test.StorageClassOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-storage-class",
+				},
+				Provisioner:       aws.String("ebs.csi.aws.com"),
+				VolumeBindingMode: lo.ToPtr(storagev1.VolumeBindingWaitForFirstConsumer),
+			})
+			ExpectSetEBSDriverLimit(1)
```
Comment (jonathan-innis) [inline test/suites/integration/storage_test.go:120]: Just curious: Does this need to be set in each test case or just in the ones where we care about the volume limit?

File: test/suites/integration/storage_test.go (hunk 2 above)
Comment (jmdeal) [reply test/suites/integration/storage_test.go:93]: We could do that but I don't think it's necessary to test the topology logic. The PVC needs to bind to the PV before it can schedule, once it binds the pod gains the PVs topology constraints. IIUC these constraints would cause the pod to fail to schedule if Karpenter launched a node in the wrong AZ.

File: test/suites/integration/storage_test.go (hunk 1 above)
Comment (jmdeal) [reply test/suites/integration/storage_test.go:57]: Do you mean with an ephemeral persistent volume (that makes sense to add). I don't think a dynamic PV with a non-existent storage class makes sense to test because it would fail to schedule. This should be covered in the functional testing.

**Review by jonathan-innis — APPROVED** [review]
Awesome work here 🎉

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀
